sdks/python/apache_beam/transforms/trigger.py

```diff
@@ -1162,12 +1173,33 @@ def process_timer(self, window_id, unused_name, time_domain, timestamp,
         if self.trigger_fn.should_fire(time_domain, timestamp,
                                        window, context):
           finished = self.trigger_fn.on_fire(timestamp, window, context)
-          yield self._output(window, finished, state)
+          yield self._output(window, finished, state, timestamp,
+                             time_domain == TimeDomain.WATERMARK)
     else:
       raise Exception('Unexpected time domain: %s' % time_domain)
 
-  def _output(self, window, finished, state):
+  def _output(self, window, finished, state, watermark, maybe_ontime):
     """Output window and clean up if appropriate."""
+    index = state.get_state(window, self.INDEX)
+    state.add_state(window, self.INDEX, 1)
+    if watermark <= window.max_timestamp():
+      nonspeculative_index = -1
+      timing = windowed_value.PaneInfoTiming.EARLY
+      if state.get_state(window, self.NONSPECULATIVE_INDEX):
+        logging.warning('Watermark moved backwards in time.')
```

Comment:
Session window max_timestamp can move forward which can result in this log.

Reply:
Good point. Added that this can happen if data moves the window max timestamp forward (which can only happen here if the data is late).